### PR TITLE
[RFC] Make SparseTensorRef the same as Tensor

### PR DESCRIPTION
**Pull request by yf225**
As part of the Variable/Tensor merge, we will make VariableType `unpack()` call `var.tensor_data()` to get the equivalent Tensor of a Variable. Under the hood, `var.tensor_data()` return a temporary object which is a shallow copy of `var`. This causes problem with the sparse tensor version of `unpack()`:
```cpp
// torch/csrc/autograd/VariableTypeManual.cpp
SparseTensorRef VariableType::unpack(SparseTensorRef t, const char * name, int pos) {
  return SparseTensorRef(checked_cast_variable(t.tref, name, pos).tensor_data());
}
```
```cpp
// aten/src/ATen/core/SparseTensorRef.h
class Tensor;
struct SparseTensorRef {
  explicit SparseTensorRef(const Tensor& t): tref(t) {}
  const Tensor& tref;
};
```
In the first code block for `unpack()`, `.tensor_data()` returns a temporary object, but since SparseTensorRef only stores a const reference of Tensor `tref` (which doesn't extend the lifetime of a temporary according to https://stackoverflow.com/questions/2784262/does-a-const-reference-class-member-prolong-the-life-of-a-temporary), after `unpack()` is done, the tensor referenced by SparseTensorRef `tref` is gone, and accessing the returned value of `unpack()` will result in error. One example of such failure is in `VariableType::sparse_mask`:

```cpp
Tensor VariableType::sparse_mask(const Tensor & self, SparseTensorRef mask) const {
  ...
  auto mask_ = unpack(mask, "mask", 1);
  std::cout << "mask_: " << mask_.tref << std::endl;  // This segfaults! Because `tref` can't hold a temporary object
  ...
}
```

One solution to this is to make SparseTensorRef the same as Tensor, and allow sparse tensors to be passed around using the Tensor class. The reasoning is that since Tensor is just wrapper of TensorImpl, and copying a Tensor doesn't deep copy the underlying TensorImpl, there should be no need to have another SparseTensorRef wrapper class that wraps the Tensor, which will also solve the problem we encounter here.

### CODE REVIEW

**Comment by gchanan** [comment]
paging @ezyang as the SparseTensorRef expert.

**Comment by ezyang** [comment]
A few comments:

- I agree with the diagnosis of the problem. Previously the Tensor object was preexisting and no problem; now you're unwrapping and therefore you got problems.
- However, I am a bit confused why unwrapping is happening at all in the first place. I was under the impression that we weren't supposed to do any unwrapping (as that implies we have to allocate new Tensor objects on every function call; wasteful!) Instead, we were supposed to just toggle a thread local but pass along the original thing. Is this just an intermediate state problem, and in the terminal state there's no problem?
- I think you should really define `SparseTensorRef = const Tensor&`. If the diff compiles in this form, I think it is good to go. The thing I am worried is if we have some function somewhere that looks like `f(SparseTensorRef x); f(const Tensor& x);` and after your change we'd have an overload for `Tensor` and `const Tensor&` which would be hella confusing. My *hope* is that, actually, we don't have any such overloads, but as the patch is currently written I'm not 100% sure this is the case. I could do a grep through the codebase to try to find out, if you want me to.

File: aten/src/ATen/function_wrapper.py [hunk 1]
```diff
@@ -291,7 +291,7 @@ def __init__(self, reason):
     'THSTensor*':
         CodeTemplate(
             'checked_tensor_unwrap('
-            '${arg_name}.tref,"${arg_name}",${arg_pos},false, '
+            '${arg_name},"${arg_name}",${arg_pos},false, '
```
Comment (gchanan) [inline aten/src/ATen/function_wrapper.py]: I'm guess this is dead code, because we don't have `THSTensor` anymore.

Comment (yf225) [reply aten/src/ATen/function_wrapper.py]: I removed the `'THSTensor*'` case from `CHECKED_CAST`.

**Comment by yf225** [comment]
> However, I am a bit confused why unwrapping is happening at all in the first place. I was under the impression that we weren't supposed to do any unwrapping (as that implies we have to allocate new Tensor objects on every function call; wasteful!) Instead, we were supposed to just toggle a thread local but pass along the original thing. Is this just an intermediate state problem, and in the terminal state there's no problem?

Yes this is just an intermediate state problem, because for now we want to preserve the invariant that ATen doesn't deal with Variables and only with Tensors. After we have the Variable::Impl removal and the optimization of not having AutogradMeta when a Tensor doesn't require grad, the distinction between a Variable and a Tensor doesn't exist anymore, and at that point we can remove the shallow-copying in `unpack()` to make the VariableType path faster.

> I think you should really define SparseTensorRef = const Tensor&. If the diff compiles in this form, I think it is good to go. The thing I am worried is if we have some function somewhere that looks like f(SparseTensorRef x); f(const Tensor& x); and after your change we'd have an overload for Tensor and const Tensor& which would be hella confusing. My hope is that, actually, we don't have any such overloads, but as the patch is currently written I'm not 100% sure this is the case. I could do a grep through the codebase to try to find out, if you want me to.

I tried defining `SparseTensorRef = const Tensor&` and it seems to compile, so I think we are good here. :)

**Pull request closed without merging**